File: 백준/Silver/1012. 유기농 배추/유기농 배추.java

import java.util.Scanner;

public class Main {

	static int m;
	static int n;

	public static void main(String[] args) {

		Scanner sc = new Scanner(System.in);

		int t = sc.nextInt();
		int tt = 1;

		while (tt <= t) {
			m = sc.nextInt(); // 가로
			n = sc.nextInt(); // 세로
			int k = sc.nextInt(); // 배추 갯수

			int[][] grid = new int[n][m];

			for (int i = 0; i < k; i++) {
				int a = sc.nextInt();
				int b = sc.nextInt();
				grid[b][a] = 1;
			}

//		for (int i = 0; i < n; i++) {
//			for (int j = 0; j < m; j++) {
//				System.out.print(grid[i][j] + " ");
//			}
//			System.out.println();
//		}

			int cnt = 0;
			for (int i = 0; i < n; i++) {
				for (int j = 0; j < m; j++) {
					if (grid[i][j] == 1) {
						cnt++;
						dfs(grid, i, j);
					}
				}
			}
			System.out.println(cnt);
			tt++;
		}

	}

	private static void dfs(int[][] grid, int i, int j) {
		int[] row = { 1, -1, 0, 0 };
		int[] col = { 0, 0, 1, -1 };

		grid[i][j] = 0;

		for (int k = 0; k < 4; k++) {
			int nr = i + row[k];
			int nc = j + col[k];

			if (nr >= 0 && nr < n && nc >= 0 && nc < m && grid[nr][nc] == 1) {
				dfs(grid, nr, nc);
			}
		}

	}
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys
sys.setrecursionlimit(10**5) # 애도 까먹음
T = int(input())
for tc in range(T):
    '''
    군집? 구하기
    r,c 가 반대네
    '''
    m, n, k = map(int, input().split())
    grid = [[0] * m for i in range(n)]

    for i in range(k):
        c, r = map(int, input().split())
        grid[r][c] = 1

    visited = [[0] * m for i in range(n)]

    row = [-1, 1, 0, 0]
    col = [0, 0, 1, -1]


    def dfs(r, c):
        visited[r][c] = 1
        for k in range(4):
            nr = r + row[k]
            nc = c + col[k]
            if (0 <= nr < n and 0 <= nc < m and visited[nr][nc] == 0 and grid[nr][nc] == 1):
                dfs(nr, nc)

    ans = 0
    for i in range(n):
        for j in range(m): # n으로 해서 헤맸음
            if (grid[i][j] == 1 and visited[i][j] == 0):
                ans += 1
                dfs(i, j)
    print(ans)